rust/connlib/tunnel/src/tests.rs

```diff
@@ -0,0 +1,359 @@
+}
+
+/// The actual system-under-test.
+///
+/// [`proptest`] manipulates this using [`Transition`]s and we assert it against [`ReferenceState`].
+struct TunnelTest {
+    now: Instant,
+
+    client: ClientState,
+    gateway: GatewayState,
+
+    #[allow(dead_code)]
+    logger: DefaultGuard,
+}
+
+/// The reference state machine of the tunnel.
+///
+/// This is the "expected" part of our test.
+#[derive(Clone, Debug)]
+struct ReferenceState {
+    now: Instant,
+    client_priv_key: [u8; 32],
+    gateway_priv_key: [u8; 32],
+
+    /// Which resources the clients is aware of.
+    client_resources: IpNetworkTable<ResourceDescriptionCidr>,
+    /// Cache for resource IPs.
+    ip4_resource_ips: Vec<Ipv4Addr>,
+    ip6_resource_ips: Vec<Ipv6Addr>,
+
+    last_connection_intents: HashMap<ResourceId, Instant>,
+
+    /// New events that we expect to be emitted.
+    client_new_events: Vec<(Instant, ClientEvent)>,
+}
+
+impl StateMachineTest for TunnelTest {
+    type SystemUnderTest = Self;
+    type Reference = ReferenceState;
+
+    // Initialize the system under test from our reference state.
+    fn init_test(
+        ref_state: &<Self::Reference as ReferenceStateMachine>::State,
+    ) -> Self::SystemUnderTest {
+        Self {
+            now: ref_state.now,
+            client: ClientState::new(StaticSecret::from(ref_state.client_priv_key)),
+            gateway: GatewayState::new(StaticSecret::from(ref_state.gateway_priv_key)),
+            logger: tracing_subscriber::fmt()
+                .with_test_writer()
+                .with_env_filter("debug")
+                .finish()
+                .set_default(),
+        }
+    }
+
+    // Apply a generated state transition to our system under test and assert against the reference state machine.
+    fn apply(
+        mut state: Self::SystemUnderTest,
+        ref_state: &<Self::Reference as ReferenceStateMachine>::State,
+        transition: <Self::Reference as ReferenceStateMachine>::Transition,
+    ) -> Self::SystemUnderTest {
+        match transition {
+            Transition::AddCidrResource(r) => {
+                state.client.add_resources(&[ResourceDescription::Cidr(r)]);
+            }
+            Transition::SendICMPPacketToResource { src, dst } => {
+                let _maybe_transmit = state
+                    .client
+                    .encapsulate(icmp_request_packet(src, dst), state.now);
+
+                // TODO: Handle transmit (send to relay / gateway)
+            }
+            Transition::Tick { millis } => {
+                state.now += Duration::from_millis(millis);
+                state.client.handle_timeout(state.now);
+                state.gateway.handle_timeout(state.now);
+            }
+        };
+
+        for (time, expected_event) in &ref_state.client_new_events {
+            assert_eq!(time, &state.now);
+            assert_eq!(expected_event, &state.client.poll_event().unwrap());
+        }
```

Comment:
So far, these are the only assertions. The properties that we test so far are:

- ICMPs to resources generate connection intents every 2s
- ICMPs to non-resources have no effect